test/common/upstream/health_check_fuzz.cc

```diff
@@ -221,6 +254,214 @@ void TcpHealthCheckFuzz::raiseEvent(const Network::ConnectionEvent& event_type,
+  ENVOY_LOG_MISC(trace, "Created Test Grpc Health Checker");
+}
+
+void GrpcHealthCheckFuzz::initialize(test::common::upstream::HealthCheckTestCase input) {
+  allocGrpcHealthCheckerFromProto(input.health_check_config());
+  cluster_->prioritySet().getMockHostSet(0)->hosts_ = {
+      makeTestHost(cluster_->info_, "tcp://127.0.0.1:80")};
+  expectSessionCreate();
+  expectStreamCreate(0);
+  health_checker_->start();
+  ON_CALL(runtime_.snapshot_, getInteger("health_check.min_interval", _))
+      .WillByDefault(testing::Return(45000));
+
+  if (DurationUtil::durationToMilliseconds(input.health_check_config().initial_jitter()) != 0) {
+    test_sessions_[0]->interval_timer_->invokeCallback();
+  }
+
+  reuse_connection_ =
+      PROTOBUF_GET_WRAPPED_OR_DEFAULT(input.health_check_config(), reuse_connection, true);
+}
+
+// Logic from respondResponseSpec() in unit tests
+void GrpcHealthCheckFuzz::respond(test::common::upstream::GrpcRespond grpc_respond) {
+  if (!test_sessions_[0]->timeout_timer_->enabled_) {
+    ENVOY_LOG_MISC(trace, "Timeout timer is disabled. Skipping response.");
+    return;
+  }
+  // These booleans help figure out when to end the stream
+  const bool has_data = grpc_respond.has_grpc_respond_bytes();
+  // Didn't hardcode grpc-status to fully explore search space provided by codecs.
+
+  // If the fuzzing engine generates a grpc_respond_trailers message, there is a validation
+  // that trailers (test.fuzz.Headers) must be present. If it is present, that means there is
+  // trailers that will be passed to decodeTrailers(). An empty trailer map counts as having
+  // trailers.
+  const bool has_trailers = grpc_respond.has_grpc_respond_trailers();
+
+  ENVOY_LOG_MISC(trace, "Has data: {}. Has trailers: {}.", has_data, has_trailers);
+
+  bool end_stream_on_headers = !has_data && !has_trailers;
+
+  std::unique_ptr<Http::TestResponseHeaderMapImpl> response_headers =
+      std::make_unique<Http::TestResponseHeaderMapImpl>(
+          Fuzz::fromHeaders<Http::TestResponseHeaderMapImpl>(
+              grpc_respond.grpc_respond_headers().headers(), {}, {}));
+
+  response_headers->setStatus(grpc_respond.grpc_respond_headers().status());
+
+  ENVOY_LOG_MISC(trace, "Responded headers {}", *response_headers.get());
+  test_sessions_[0]->stream_response_callbacks_->decodeHeaders(std::move(response_headers),
+                                                               end_stream_on_headers);
+
+  // If the interval timer is enabled, that means that the rpc is complete, as decodeHeaders hit a
+  // certain branch that called onRpcComplete(), logically representing a completed rpc call. Thus,
+  // skip the next responses until explicility invoking interval timer as cleanup.
+  if (has_data && !test_sessions_[0]->interval_timer_->enabled_) {
+    std::vector<std::vector<uint8_t>> bufferList;
+
+    switch (grpc_respond.grpc_respond_bytes().grpc_respond_bytes_selector_case()) {
```

Comment:
final comment just on this switch statement -- did makeBufferList end up dropped? I liked how that worked! I think it should apply to both cases though (as in, take grpc_respond_bytes(), switch on type, construct a buffer list for each case), and this logic is just executing the for loop that decodes data, common to both

Reply:
Yeah deleted makeBufferList!

Reply:
It's not common to both. Serialized responses get split up evenly, while the other bufferList is repeated arbitrary length bytes.

Reply:
Example on line 343: access only the first element of buffer list, and moves pointer to right i :).

Reply:
Line 368: iterates through the bufferList itself, not memory within bufferList[0] like serialized responses.

Reply:
Hmmm we can chat offline about this -- I'm still failing to see why `respond` has to know about that --
i.e. makeBuffers can handle both of those cases, and return a buffer with chunks (may be evenly split, may be arbitrary length) to `respond`, and `respond` can iterate through the chunks calling decodeData.